### PR TITLE
Add support for integer arguments

### PR DESCRIPTION
**Pull request by jsullivan3**
Integer arguments seem to have been an oversight.  This PR adds support for processing integer arguments, as well as tests to verify the correct operation of integer argument processing.

### CODE REVIEW

File: argument.go [hunk 1]
```diff
@@ -188,6 +202,8 @@ func (o *arg) usage() string {
 	switch o.result.(type) {
 	case *bool:
 		break
+	case *int:
+		result = result + " <value>"
```
Comment (akamensky) [inline argument.go]: Technically `<value>` is okay, but I would imagine may be not clear enough for whomever reads help message. So maybe `<int>` would be good? Not necessary to change, just a thought.

File: argument.go [hunk 2]
```diff
@@ -121,6 +122,19 @@ func (o *arg) parse(args []string) error {
...
+	case *int:
+		if len(args) < 1 {
+			return fmt.Errorf("[%s] must be followed by an integer", o.name())
+		}
+		if len(args) > 1 {
+			return fmt.Errorf("[%s] followed by too many arguments", o.name())
+		}
+		val, err := strconv.Atoi(args[0])
+		if err != nil {
+			return fmt.Errorf("[%s] must be followed by an integer", o.name())
```
Comment (akamensky) [inline argument.go]: Hm, not sure whether error message in this case is good. The arg is already followed by something, which is not an integer, not sure if can improve this. Just something to think about.

File: argparse_test.go [hunk 3]
```diff
@@ -792,6 +1035,6 @@ func TestSelectorDefaultValueFail(t *testing.T) {
 
 	// Should pass on failure
 	if err == nil || err.Error() != "cannot use default type [bool] as type [string]" {
-		t.Errorf("Test %s failed: expected error [%s], got error [%s]", t.Name(), "cannot use default type [bool] as type [string]", err.Error())
+		t.Errorf("Test %s failed: expected error [%s], got error [%+v]", t.Name(), "cannot use default type [bool] as type [string]", err)
```
Comment (akamensky) [inline argparse_test.go]: Unnecessary change in context of this PR

File: argparse_test.go [hunk 4]
```diff
@@ -756,7 +999,7 @@ func TestListDefaultValueFail(t *testing.T) {
 
 	// Should pass on failure
 	if err == nil || err.Error() != "cannot use default type [bool] as type [[]string]" {
-		t.Errorf("Test %s failed: expected error [%s], got error [%s]", t.Name(), "cannot use default type [bool] as type [[]string]", err.Error())
+		t.Errorf("Test %s failed: expected error [%s], got error [%+v]", t.Name(), "cannot use default type [bool] as type [[]string]", err)
```
Comment (akamensky) [inline argparse_test.go]: Unnecessary change in context of this PR

File: argparse_test.go [hunk 5]
```diff
@@ -719,7 +962,7 @@ func TestFileDefaultValueFail(t *testing.T) {
 
 	err = p.Parse(testArgs)
 	if err == nil || err.Error() != "cannot use default type [bool] as type [string]" {
-		t.Errorf("Test %s failed: expected error [%s], got error [%s]", t.Name(), "cannot use default type [bool] as type [string]", err.Error())
+		t.Errorf("Test %s failed: expected error [%s], got error [%+v]", t.Name(), "cannot use default type [bool] as type [string]", err)
```
Comment (akamensky) [inline argparse_test.go]: Unnecessary change in context of this PR

File: argparse_test.go [hunk 6]
```diff
@@ -669,7 +876,43 @@ func TestStringDefaultValueFail(t *testing.T) {
 
 	// Should pass on failure
 	if err == nil || err.Error() != "cannot use default type [bool] as type [string]" {
-		t.Errorf("Test %s failed: expected error [%s], got error [%s]", t.Name(), "cannot use default type [bool] as type [string]", err.Error())
+		t.Errorf("Test %s failed: expected error [%s], got error [%+v]", t.Name(), "cannot use default type [bool] as type [string]", err)
```
Comment (akamensky) [inline argparse_test.go]: Unnecessary change in context of this PR

File: argparse_test.go [hunk 7]
```diff
@@ -633,7 +840,7 @@ func TestFlagDefaultValueFail(t *testing.T) {
 
 	// Should pass on failure
 	if err == nil || err.Error() != "cannot use default type [string] as type [bool]" {
-		t.Errorf("Test %s failed: expected error [%s], got error [%s]", t.Name(), "cannot use default type [string] as type [bool]", err.Error())
+		t.Errorf("Test %s failed: expected error [%s], got error [%+v]", t.Name(), "cannot use default type [string] as type [bool]", err)
```
Comment (akamensky) [inline argparse_test.go]: Unnecessary change in context of this PR

File: argument.go (hunk 1 above)
Comment (jsullivan3) [reply argument.go]: Makes sense to me - thanks!  I'll take a look at the Python `argparse` module and use it as a reference.  If it uses `<int>` or something similar, I'll go with that change for consistency.

File: argument.go (hunk 2 above)
Comment (jsullivan3) [reply argument.go]: Another good point.  I used this wording since to get to this error, the value following the argument is not an integer otherwise `strconv.Atoi()` would not have returned an error...so the message is intended to emphasize that the parameter to the argument needs to be an _integer_.  I can't think of any improvements, though I'm very open to suggestions.

File: argparse_test.go (hunk 3 above)
Comment (jsullivan3) [reply argparse_test.go]: Well....it's technically wrong, and I had to fix it in the test I created using this code as a base.  If control enters this block because `err == nil`, the dereference of `err.Error()` causes the test to segfault.  I can remove these fixes if you wish, though I found them to be useful since I triggered the analogous error in the tests I added and found the message to be more helpful than the segfault... :)

File: argument.go (hunk 1 above)
Comment (akamensky) [reply argument.go]: I am not sure that Python module does even have such message differences. This module inspired by it, but does not really copy it. With help messages I went after what's common on _classic_ CLI tools, which of course may differ one to another. I guess just need to use common sense here. If you (as a dev) feel it will be clear enough value should be numeric, then I am good with that.

File: argparse_test.go (hunk 3 above)
Comment (akamensky) [reply argparse_test.go]: Yes, it might be wrong, but suggest to send it as separate PR in this case ;) One PR -- one [logical] change, is what I strive for.

**Comment by akamensky** [comment]
Also before I can merge it, can you please squash all commits to one. That would be very helpful!

File: argument.go (hunk 1 above)
Comment (jsullivan3) [reply argument.go]: Turns out it does not, but the error message that gets presented when a non-integer value is passed to an integer parameter is more clear and references `int`.  I'll modify this code to hopefully make it a little more clear and readable.  Thanks again!

File: argparse_test.go (hunk 3 above)
Comment (jsullivan3) [reply argparse_test.go]: Completely understood - I tend to fight to balance between "one PR, one logical change" and "fix what's broken."  I will gladly create a second PR to address these (admittedly rare!) test failures.  Thanks!

**Comment by jsullivan3** [comment]
Will squash as requested - not a problem.  And thank you for this fantastic Go package.  It allowed me to greatly improve the usability of the CLIs that I'm writing.

**Comment by jsullivan3** [comment]
I think the current squashed commit covers all the changes requested - please let me know if you have any other feedback and I'll gladly implement it.  Thanks again for your time in reviewing this change.

**Comment by akamensky** [comment]
And thank you for this PR!

**Comment by jsullivan3** [comment]
Awesome - you are very welcome!
